src/qoa_decoder.h

```diff
@@ -0,0 +1,524 @@
+	for (int i = 0; i < QOA_LMS_LEN; i++) {
+		lms->weights[i] += lms->history[i] < 0 ? -delta : delta;
+	}
+
+	for (int i = 0; i < QOA_LMS_LEN-1; i++) {
+		lms->history[i] = lms->history[i+1];
+	}
+	lms->history[QOA_LMS_LEN-1] = sample;
+}
+
+
+/* qoa_div() implements a rounding division, but avoids rounding to zero for
+small numbers. E.g. 0.1 will be rounded to 1. Note that 0 itself still
+returns as 0, which is handled in the qoa_quant_tab[].
+qoa_div() takes an index into the .16 fixed point qoa_reciprocal_tab as an
+argument, so it can do the division with a cheaper integer multiplication. */
+
+static inline int qoa_div(int v, int scalefactor) {
+	int reciprocal = qoa_reciprocal_tab[scalefactor];
+	int n = (v * reciprocal + (1 << 15)) >> 16;
+	n = n + ((v > 0) - (v < 0)) - ((n > 0) - (n < 0)); /* round away from 0 */
+	return n;
+}
+
+static inline int qoa_clamp(int v, int min, int max) {
+	if (v < min) { return min; }
+	if (v > max) { return max; }
+	return v;
+}
+
+/* This specialized clamp function for the signed 16 bit range improves decode
+performance quite a bit. The extra if() statement works nicely with the CPUs
+branch prediction as this branch is rarely taken. */
+
+static inline int qoa_clamp_s16(int v) {
+	if ((unsigned int)(v + 32768) > 65535) {
+		if (v < -32768) { return -32768; }
+		if (v >  32767) { return  32767; }
+	}
+	return v;
+}
+
+static inline qoa_uint64_t qoa_read_u64(const unsigned char *bytes, unsigned int *p) {
+	bytes += *p;
+	*p += 8;
+	return
+		((qoa_uint64_t)(bytes[0]) << 56) | ((qoa_uint64_t)(bytes[1]) << 48) |
+		((qoa_uint64_t)(bytes[2]) << 40) | ((qoa_uint64_t)(bytes[3]) << 32) |
+		((qoa_uint64_t)(bytes[4]) << 24) | ((qoa_uint64_t)(bytes[5]) << 16) |
+		((qoa_uint64_t)(bytes[6]) <<  8) | ((qoa_uint64_t)(bytes[7]) <<  0);
+}
+
+/* -----------------------------------------------------------------------------
+	Decoder */
+
+static unsigned int qoa_decode_header(qoa_data *data) {
+	unsigned int p = 0;
+	if (data->size < QOA_MIN_FILESIZE) {
+		return 0;
+	}
+
+	/* Read the file header, verify the magic number ('qoaf') and read the
+	total number of samples. */
+	qoa_uint64_t file_header = qoa_read_u64(data->bytes, &p);
+
+	if ((file_header >> 32) != QOA_MAGIC) {
+		return 0;
+	}
+
+	data->qoa.samples = file_header & 0xffffffff;
+	if (!data->qoa.samples) {
+		return 0;
+	}
+
+	/* Peek into the first frame header to get the number of channels and
+	the samplerate. */
+	qoa_uint64_t frame_header = qoa_read_u64(data->bytes, &p);
+	data->qoa.channels   = (frame_header >> 56) & 0x0000ff;
+	data->qoa.samplerate = (frame_header >> 32) & 0xffffff;
+	data->samples_per_channel_per_frame = (frame_header >> 16) & 0x00ffff;
+
+	if (data->qoa.channels == 0 || data->qoa.channels == 0 || data->qoa.samplerate == 0) {
+		return 0;
+	}
+
+	return 8;
+}
+
+qoa *qoa_open_from_memory(unsigned char *bytes, unsigned int size, int free_on_close)
+{
+	qoa_data *data = (qoa_data*) FAudio_malloc(sizeof(qoa_data));
```

Comment:
Similar to stb it might be easier to macro these stdlib functions ahead of the include, so we're as close to upstream as possible